File: Grind 75 in Java/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int l = 0;
        int max = 0;

        for (int r = 1; r < prices.length; r++) {
            if (prices[l] < prices[r]) {
                int profit = prices[r] - prices[l];
                max = Math.max(max, profit);
            } else {
                l = r;
            }
        }

        return max;
    }
}

File: leetcode in kotlin/0121-best-time-to-buy-and-sell-stock.kt
class Solution { 
    fun maxProfit(prices: IntArray): Int {
        var r = 1; var l = 0; var max = 0  
        while (r < prices.size){
            if (prices[l] < prices[r]){
                val profit = prices[r] - prices[l]
                max = maxOf(max, profit)
            }else
                l = r        
            r++
        }
    return max
    }  
}

// Kadane's Algorithm
class Solution {
    fun maxProfit(prices: IntArray): Int {
        var cur = 0
        var maxSoFar = 0

        for (i in 1 until prices.size) {
            cur += prices[i] - prices[i - 1]
            cur = maxOf(0, cur)
            maxSoFar = maxOf(maxSoFar, cur)
        }

        return maxSoFar
    }
}
